File: src/principal/Amortizacao.java
package principal;
import java.math.RoundingMode;
import java.text.DecimalFormat;

public class Amortizacao {
    double periodo, saldoatual, taxajuros;
    
    // Construtor inicializador
    public Amortizacao (double pr, double sa, double txj) {
        periodo = pr;
        saldoatual = sa;
        taxajuros = txj;
    }

    // Construtor cópia
    public Amortizacao(Amortizacao A) {
        this.periodo = A.periodo;
        this.saldoatual = A.saldoatual;
        this.taxajuros = A.taxajuros;
    }
        
    // Destrutor de classe
    public void finalize() {}

    // Calculo de expoente
    static double potencia(double n, double power) {
        double res = 1;
        for(double i = 1; i <= power; i++) {
            res = res*n;
        }
        return res;
    }

    // Exibição dos elementos da tabela S.A.C/S.A.F/S.A.M/S.A.Alemão convertidos em Real e Decimal
    public static void exibirTAB(int valor0, double valor1, double valor2, double valor3, double valor4) {
    	
        DecimalFormat decimalFormat = new DecimalFormat("R$#,##0.00");
        decimalFormat.setRoundingMode(RoundingMode.DOWN);
        System.out.println("Período: " + valor0 + "\tSaldo atual:\t" + decimalFormat.format(valor1)
        +" \tAmortização:\t"+ decimalFormat.format(valor2)+"\tJuros: "+ decimalFormat.format(valor3)
        +"\t\tPrestação:\t" + decimalFormat.format(valor4));
    }

    // Exibição os elementos do periodo 0 das tabelas S.A.C/S.A.F/S.A.M/S.A.Alemão convertidos em Real e Decimal
    public static void exibir0(int valor0, double valor1, double valor2, double valor3, double valor4) {
    	
        DecimalFormat decimalFormat = new DecimalFormat("R$#,##0.00");
        decimalFormat.setRoundingMode(RoundingMode.DOWN);
        System.out.println("Período: " + valor0 + "\tSaldo atual:\t" + decimalFormat.format(valor1)
        +"\tAmortização:\t"+decimalFormat.format(valor2)+"\t\tJuros:\t"+decimalFormat.format(valor3)
        +"\t\tPrestação:\t" + decimalFormat.format(valor4));
    }
    
    // Exibição os elementos do periodo 0 do fundo de amortizaçao da tabela S.A.A convertidos em Real e Decimal
    public static void exibir1(int valor0, double valor1, double valor2, double valor3) {
        DecimalFormat decimalFormat = new DecimalFormat("R$#,##0.00");
        decimalFormat.setRoundingMode(RoundingMode.DOWN);
        System.out.println("Período: " + valor0 + "\tDepósito:\t" + decimalFormat.format(valor1)
        +"\tJuros:\t"+decimalFormat.format(valor2)+"\t\tMontante:\t"+decimalFormat.format(valor3));
    }

    // Exibição dos elementos da planilha financeira da tabela S.A.A convertidos em Real e Decimal
    public static void exibirSAA1(int valor0, double valor1, double valor2, double valor3, double valor4) {
    	
        DecimalFormat decimalFormat = new DecimalFormat("R$#,##0.00");
        decimalFormat.setRoundingMode(RoundingMode.DOWN);
        System.out.println("Período: " + valor0 + "\tSaldo atual:\t" + decimalFormat.format(valor1) 
        + "\tAmortização:\t" + decimalFormat.format(valor2) + "\tJuros:\t" + decimalFormat.format(valor3) 
        + "\tPrestação:\t" + decimalFormat.format(valor4));   
    }

    // Exibição os elementos do fundo de amortizaçao da tabela S.A.A convertidos em Real e Decimal
    public static void exibirSAA2(int valor0, double valor1, double valor2, double valor3) {
        DecimalFormat decimalFormat = new DecimalFormat("R$#,##0.00");
        decimalFormat.setRoundingMode(RoundingMode.DOWN);
        System.out.println("Período: " + valor0 + "\tDepósito:\t" + decimalFormat.format(valor1) + "\tJuros:\t" + decimalFormat.format(valor2) 
        + "\tMontante:\t" + decimalFormat.format(valor3));
    }

    // Método S.A.C - Sistema de Amortização Constante
    public void SAC() {
        DecimalFormat decimalFormat = new DecimalFormat("R$#,##0.00");
        decimalFormat.setRoundingMode(RoundingMode.DOWN);
        double amortizacao, juros, prestacao, tam=0, tj=0, tp=0;
        amortizacao = saldoatual/periodo;
        for (int i = 1; i<=periodo; i++) {
            juros = saldoatual*taxajuros;
            prestacao = amortizacao + juros;
            saldoatual -= amortizacao;
            tam = tam + amortizacao;
            tj = tj + juros;
            tp = tp + prestacao;
            exibirTAB(i, saldoatual, amortizacao, juros, prestacao);
        }
        System.out.println("\nTotalizações - \tAmortização: "+decimalFormat.format(tam)+"\tJuros: "+decimalFormat.format(tj)
        +"\tPrestacao: "+decimalFormat.format(tp));
    }

    // Método S.A.F - Sistema de Amortização Francês - Tabela Price
    public void SAF() {
        DecimalFormat decimalFormat = new DecimalFormat("R$#,##0.00");
        decimalFormat.setRoundingMode(RoundingMode.DOWN);
        double coef = ((potencia((1+taxajuros),periodo))*taxajuros)/((potencia((1+taxajuros),periodo))-1);
        double amortizacao=0, juros=0, saldodevedor, prestacao=0, tam=0, tj=0, tp=0;
        int i = 0;
        exibir0(i, saldoatual, amortizacao, juros, prestacao);
        prestacao = saldoatual*coef;
        
        for(i = 1; i<=periodo; i++) {
            juros = saldoatual*taxajuros;
            amortizacao = prestacao - juros;
            saldodevedor = saldoatual + juros;
            saldoatual = saldodevedor - prestacao;
            tam += amortizacao;
            tj += juros;
            tp += prestacao;
            exibirTAB(i, saldoatual, amortizacao, juros, prestacao);
        }
        System.out.println("Totalizações - \t\t\t\t\tAmortização:\t"+decimalFormat.format(tam)+"\tJuros: "+decimalFormat.format(tj)
        +"\t\tPrestacao:\t"+decimalFormat.format(tp));
    }

    
    // Método S.A.M - Sistema de Amortização Misto
    public void SAM() {
        DecimalFormat decimalFormat = new DecimalFormat("R$#,##0.00");
        decimalFormat.setRoundingMode(RoundingMode.DOWN);
        double coef = ((potencia((1+taxajuros),periodo))*taxajuros)/((potencia((1+taxajuros),periodo))-1);
        double amortizacao=0, amortizacaoSAC, juros=0, prestacaoSAF=0, prestacaoSAC, prestacaoSAM=0, saldoatualSAC=saldoatual, tam=0, tj=0, tp=0;
        int i = 0;
        exibir0(i, saldoatual, amortizacao, juros, prestacaoSAM);
        prestacaoSAF = saldoatual*coef;
        amortizacaoSAC=(saldoatual/periodo);
        for(i = 1; i<=periodo; i++) {
            prestacaoSAC = amortizacaoSAC + (saldoatualSAC*taxajuros);
            juros = saldoatual*taxajuros;
            saldoatualSAC -= amortizacaoSAC;
            prestacaoSAM = (prestacaoSAC + prestacaoSAF)/2;
            amortizacao = prestacaoSAM - juros;
            saldoatual -= amortizacao;
            tam += amortizacao;
            tj += juros;
            tp += prestacaoSAM;
            exibirTAB(i, saldoatual, amortizacao, juros, prestacaoSAM);
        }
        System.out.println("Totalizações - \t\t\t\t\tAmortização:\t"+decimalFormat.format(tam)+"\tJuros:\t"+decimalFormat.format(tj)
        +"\tPrestacao:\t"+ decimalFormat.format(tp));
    }

    // Método S.A.Alemão - Sistema de Amortização Alemão
    public void SAALM() {
        DecimalFormat decimalFormat = new DecimalFormat("R$#,##0.00");
        decimalFormat.setRoundingMode(RoundingMode.DOWN);
        double amortizacaok=0, amortizacao1, juros, prestacao=0, tam=0, tj=0, tp=0;
        int i = 0;
        juros = taxajuros*saldoatual;
        exibirTAB(i, saldoatual, amortizacaok, juros, prestacao);
        prestacao = (saldoatual*taxajuros)/(1-potencia((1-taxajuros), periodo));
        amortizacao1 = prestacao * potencia((1-taxajuros),periodo-1);
        amortizacaok = amortizacao1;
        juros = prestacao - amortizacaok;
        saldoatual -= amortizacaok;
        i = 1;
        exibirTAB(i, saldoatual, amortizacaok, juros, prestacao);
        for(i = 2; i<=periodo; i++) {
            amortizacaok = amortizacaok/(1-taxajuros);
            juros = prestacao - amortizacaok;
            saldoatual -= amortizacaok;
            if (i == periodo) {
                amortizacaok = 0;
                juros = 0;
            }
            tam += amortizacaok;
            tj += juros;
            tp += prestacao;
            exibirTAB(i, saldoatual, amortizacaok, juros, prestacao);
        }
        System.out.println("Totalizações - \t\t\t\t\tAmortização:\t"+decimalFormat.format(tam)+"\tJuros:\t"+decimalFormat.format(tj)
        +"\tPrestacao:\t"+ decimalFormat.format(tp));
    }
    
}

File: Amortizacao.py
import os
import pandas as pd
import openpyxl

# tabela[0][0] = periodo        | Planilha financeira (S.A.C, S.A.F, S.A.M, S.ALM e S.A.A)
# tabela[0][1] = saldo atual    | Planilha financeira (S.A.C, S.A.F, S.A.M, S.ALM e S.A.A)
# tabela[0][2] = amortizacao    | Planilha financeira (S.A.C, S.A.F, S.A.M, S.ALM e S.A.A)
# tabela[0][3] = juros          | Planilha financeira (S.A.C, S.A.F, S.A.M, S.ALM e S.A.A)
# tabela[0][4] = prestacao      | Planilha financeira (S.A.C, S.A.F, S.A.M, S.ALM e S.A.A)
# tam = amortização total       | Planilha financeira (S.A.C, S.A.F, S.A.M, S.ALM e S.A.A)
# tj = juros total              | Planilha financeira (S.A.C, S.A.F, S.A.M, S.ALM e S.A.A)
# tp = prestação total          | Planilha financeira (S.A.C, S.A.F, S.A.M, S.ALM e S.A.A)
# tabelaFA[0][0] = depósito     | Fundo de Amortização (S.A.A)
# tabelaFA[0][1] = juros        | Fundo de Amortização (S.A.A)
# tabelaFA[0][2] = montante     | Fundo de Amortização (S.A.A)
# tm = montante total           | Fundo de Amortização (S.A.A)
# tjpf = juros total            | Fundo de Amortização (S.A.A)
# td = depósito total           | Fundo de Amortização (S.A.A)

# Loop infinito do programa

def padrao_de_colunas(worksheet):
    colunas = ['B', 'C', 'D', 'E', 'F', 'G']
    
    worksheet.column_dimensions['A'].width = 13
    
    for i in range(len(colunas)):
        worksheet.column_dimensions[colunas[i]].width = 30

def exportacao_xlsx(data_frames, planilhas):
    writer = pd.ExcelWriter(path, engine='openpyxl')
    for i, df in enumerate(data_frames):
        df.to_excel(writer, sheet_name=planilhas[i])
        workbook = writer.book
        worksheet = workbook[planilhas[i]]
        padrao_de_colunas(worksheet)
    writer._save()

def SAC():
    df = pd.DataFrame(columns=['Período', 'Saldo atual', 'Amortização', 'Juros', 'Prestação'])
    df.loc[0] = [dados_financiamento["periodo"], dados_financiamento["saldo"], 0, 0, 0]
    df.loc[0, 'Amortização'] = df.loc[0, 'Saldo atual'] / df.loc[0, 'Período']
        
    for i in range(1, df.loc[0, 'Período'] + 1):
        df.loc[i, 'Juros'] = df.loc[i-1, 'Saldo atual'] * dados_financiamento["taxajuros"]
        df.loc[i, 'Prestação'] = df.loc[0, 'Amortização'] + df.loc[i, 'Juros']
        df.loc[i, 'Saldo atual'] = df.loc[i-1, 'Saldo atual'] - df.loc[0, 'Amortização']
            
    tam = df['Amortização'].sum()
    tj = df['Juros'].sum()
    tp = df['Prestação'].sum()
        
    totais = {'Período': 'Total', 'Saldo atual': '', 'Amortização': tam, 'Juros': tj, 'Prestação': tp}
    df = df.append(totais, ignore_index=True)

    return df

# Método S.A.F - Sistema de Amortização Francês - Tabela Price
    
def SAF():
    df = pd.DataFrame(columns=['Período', 'Saldo atual', 'Amortização', 'Juros', 'Prestação'])
    tam, tj, tp, saldodevedor, df.loc[0, 'Saldo atual'], df.loc[0, 'Período'] = 0, 0, 0, 0, dados_financiamento["saldo"], dados_financiamento["periodo"]
    coef = (((1+dados_financiamento["taxajuros"])**df.loc[0, 'Período'])*dados_financiamento["taxajuros"])/(((1+dados_financiamento["taxajuros"])**df.loc[0, 'Período'])-1)
    df.loc[0, 'Prestação'] = df.loc[0, 'Saldo atual']*coef
    
    for i in range(1, df.loc[0, 'Período'] + 1):
        df.loc[i, 'Juros'] = df.loc[i-1, 'Saldo atual']*dados_financiamento["taxajuros"]
        df.loc[i, 'Amortização'] = df.loc[0, 'Prestação'] - df.loc[i, 'Juros']
        saldodevedor = df.loc[i-1, 'Saldo atual'] + df.loc[i, 'Juros']
        df.loc[i, 'Saldo atual'] = saldodevedor - df.loc[0, 'Prestação']
        tam += df.loc[i, 'Amortização']
        tj += df.loc[i, 'Juros']
        tp += df.loc[0, 'Prestação']
    
    totais = {'Período': 'Total', 'Saldo atual': '', 'Amortização': tam, 'Juros': tj, 'Prestação': tp}
    df = df.append(totais, ignore_index=True)
    
    return df

# Método S.A.A
    
def SAA():
    tabela[0][1], tabela[0][0], tam, tjpf, tjfa, tp, tm, td = dados_financiamento["saldo"], dados_financiamento["periodo"], 0, 0, 0, 0, 0, 0
    tabela[1][3] = tabela[0][1] * dados_financiamento["taxajuros"]
    tabela[1][4] = tabela[1][3]
    tabelaFA[1][0] = tabela[0][1] * (dados_financiamento["captacaomedia"]/(((1+dados_financiamento["captacaomedia"])**tabela[0][0]) - 1)) 
    tabelaFA[1][2] = tabelaFA[1][0]
        
    print("\nPLANILHA FINANCEIRA\n")
    
    print("Periodo: -Saldo atual: R$%.2f" % round(tabela[0][1],2), "  Amortizacao: - \tJuros: - \tPrestação: -")
    for i in range(1, tabela[0][0]+1):
        if(i==tabela[0][0]):
            tabela[i][2] = tabela[0][1]
            tabela[i][4] = tabela[i][2] + tabela[1][3]
            tp += tabela[i][4]
        else:
            tp += tabela[1][4]
        tam += tabela[i][2]
        tjpf += tabela[1][3]
        if (i!=tabela[0][0]):
            exibe_tab(i, tabela[0][1], tabela[1][2], tabela[1][3], tabela[1][4])
        else:
            print("Período: ", i, "Saldo atual: -  Amortização: R$%.2f" % round(tabela[i][2],2), "  Juros: R$%.2f" % round(tabela[1][3],2), "  Prestação: R$%.2f" % round(tabela[i][4],2))
    print("T O T A L I Z A Ç Õ E S --->", f"\t\t\t{'Amortização: R$%.2f' : ^20}" % round(tam,2), f"{'Juros: R$%.2f' : ^24}" % round(tjpf,2), f"{'Prestacao: R$%.2f' : ^10}" % round(tp, 2))
    
    print("\nFUNDO DE AMORTIZAÇÃO\n")
    
    print("Período:  - \tDepósito: - \tJuros: - \tMontante: -")
    for i in range(1, tabela[0][0]+1):
        if(i>1):
            tabelaFA[i][1] = tabelaFA[i-1][2] * dados_financiamento["captacaomedia"]
            tabelaFA[i][2] = tabelaFA[i-1][2] + tabelaFA[i][1]
        tjfa += tabelaFA[i][1]
        tm += tabelaFA[i][2]
        td += tabelaFA[1][0]
        exibe_tabFA(i, tabelaFA[1][0], tabelaFA[i][1], tabelaFA[i][2])
    print("T O T A L I Z A Ç Õ E S --->", f"\t{'Depósito: R$%.2f' : >20}" % round(td,2), f"\t{'Juros: R$%.2f' : >24}" % round(tjfa,2), f"\t{'Montante: R$%.2f' : >20}" % round(tm, 2))

# Método S.A.M - Sistema de Amortização Misto
    
def SAM():
    tabela[0][0], tabela[0][1] = dados_financiamento["periodo"], dados_financiamento["saldo"]
    print("\nPeríodo:  - \tSaldo atual: R$%.2f" % round(tabela[0][1],2), "\tAmortização:\t- \t\tJuros:\t- \t\tPrestação:\t-")
    coef = (((1+dados_financiamento["taxajuros"])**tabela[0][0])*dados_financiamento["taxajuros"])/(((1+dados_financiamento["taxajuros"])**tabela[0][0])-1)
    saldoatualSAC, prestacaoSAF, tam, tj, tp = tabela[0][1], tabela[0][1]*coef, 0, 0, 0
    amortizacaoSAC = (tabela[0][1]/tabela[0][0])
    for i in range(1, tabela[0][0] + 1):
        prestacaoSAC = amortizacaoSAC + (saldoatualSAC*dados_financiamento["taxajuros"])
        tabela[i][3] = tabela[i-1][1]*dados_financiamento["taxajuros"]
        saldoatualSAC -= amortizacaoSAC
        tabela[i][4] = (prestacaoSAC + prestacaoSAF)/2
        tabela[i][2] = tabela[i][4] - tabela[i][3]
        tabela[i][1] = tabela[i-1][1] - tabela[i][2]
        tam += tabela[i][2]
        tj += tabela[i][3]
        tp += tabela[i][4]
        if (i!=tabela[0][0]):
            exibe_tab(i, tabela[i][1], tabela[i][2], tabela[i][3], tabela[i][4])
        else:
            print("Período: ", i, f"{'Saldo atual: ' : >20}""\t-", f"\t{'Amortização: R$%.2f' : >15}" % round(tabela[i][2],2), f"\t{'Juros: R$%.2f' : >20}" % round(tabela[i][3],2), f"\t{'Prestação: R$%.2f' : >20}" % round(tabela[i][4],2))
    print("T O T A L I Z A Ç Õ E S --->", f"\t\t\t{'Amortização: R$%.2f' : ^20}" % round(tam,2), f"{'Juros: R$%.2f' : ^24}" % round(tj,2), f"{'Prestacao: R$%.2f' : ^10}" % round(tp, 2))

# Método S.A.Alemão - Sistema de Amortização Alemão
    
def SAALM():
    tabela[0][0], tabela[0][1], tam, tj, tp, amortizacaok, i = dados_financiamento["periodo"], dados_financiamento["saldo"], 0, 0, 0, 0, 0
    tabela[0][3] = dados_financiamento["taxajuros"]*tabela[0][1]
    tabela[0][4] = tabela[0][3]
    exibe_tab(i, tabela[0][1], amortizacaok, tabela[0][3], tabela[0][4])
    tabela[1][4] = (tabela[0][1]*dados_financiamento["taxajuros"])/(1-((1-dados_financiamento["taxajuros"])**tabela[0][0]))
    amortizacao1 = tabela[1][4] * (1-dados_financiamento["taxajuros"])**(tabela[0][0]-1)
    tabela[1][2] = amortizacao1
    tabela[1][3] = tabela[1][4] - tabela[1][2]
    tabela[1][1] = tabela[0][1] - tabela[1][2]
    exibe_tab(i+1, tabela[1][1], tabela[1][2], tabela[1][3], tabela[1][4])
    tp = tabela[0][4] + tabela[1][4]
    tam = tabela[1][2]
    tj = tabela[0][3]
    for i in range(2, tabela[0][0]+1):
        tabela[i][2] = tabela[i-1][2]/(1-dados_financiamento["taxajuros"])
        tabela[i][3] = tabela[1][4] - tabela[i][2]
        tabela[i][1] = tabela[i-1][1] - tabela[i][2]
        tam += tabela[i][2]
        tp += tabela[1][4]
        if (i == tabela[0][0]):
            tabela[i][3] = 0
        tj += tabela[i][3]
        if (i != tabela[0][0]):
            exibe_tab(i, tabela[i][1], tabela[i][2], tabela[i][3], tabela[1][4])
        else:
            print("Período: ", i, f"{'Saldo atual: ' : >20}""\t-", f"\t{'Amortização: R$%.2f' : >15}" % round(tabela[i][2],2), f"\t{'Juros: ' : >20}""\t-", f"\t{'Prestação: R$%.2f' : >20}" % round(tabela[1][4],2))
    print("T O T A L I Z A Ç Õ E S --->", f"\t\t\t{'Amortização: R$%.2f' : ^20}" % round(tam,2), f"{'Juros: R$%.2f' : ^24}" % round(tj,2), f"{'Prestacao: R$%.2f' : ^10}" % round(tp, 2))

    # Declaração de variáveis

dados_financiamento = {}

# Menu principal
    
print("\nOlá, bem vindo ao simulador de sistemas de amortização! Insira os dados de seu financiamento para prosseguir\n")

dados_financiamento["saldo"] = float(input("Insira o saldo a ser pago: "))
dados_financiamento["taxajuros"] = float(input("Insira a taxa de juros: "))
dados_financiamento["periodo"] = int(input("Insira o período: "))
caminho_relatorio = input('Insira o caminho para o diretório do arquivo: ')
path = f'{caminho_relatorio}/Simulações.xlsx'

tabela = [[0 for x in range(5)] for y in range(dados_financiamento["periodo"] + 1)]
tabelaFA = [[0 for x in range(3)] for y in range(dados_financiamento["periodo"] + 1)]

